### PR TITLE
Add additional PRNG

### PR DESCRIPTION
**Pull request by t-bast**
In case of catastrophic failures of the `SecureRandom` instance, we add a secondary randomness source that we mix into the random stream.

This is a somewhat weak random source and should not be used on its own, but it doesn't hurt to xor it with the output of `SecureRandom`.

The second commit should be reviewed independently: it simply adds parenthesis to functions that generate random things, since these have side-effects. It's trivial but painful to review since it touches a lot of test files.

NB: this feature make a lot of sense in https://github.com/acinq/lightning-kmp as well since mobile platforms may have more common randomness failures.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/package.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/package.scala:52]:
```suggestion
    val rand1 = {
      val buffer = new Array[Byte](length)
      secureRandom.nextBytes(buffer)
      ByteVector.view(buffer)
    }
    val rand2 = {
      val weakBuffer = new Array[Byte](length)
      weakRandom.nextBytes(weakBuffer)
      ByteVector.view(weakBuffer)
    }
    rand1 xor rand2
```

File: eclair-core/src/main/scala/fr/acinq/eclair/package.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/package.scala:61]: It made sense to put flat method calls in the package object when we just called basic methods on `secureRandom`, but now I think we should probably define a class that encapsulates all those xor calls, and hides `secureRandom` and `weakRandom` which are not supposed to be called directly.

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Autoprobe.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Autoprobe.scala:99]: We should define a `randomInt(bound: Int)`.

File: eclair-core/src/main/scala/fr/acinq/eclair/package.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/package.scala:42]: We should find a way to not make it possible to access `weakRandom.nextBytes` so easily. I understand that `weakRandom.addEntropy()` needs to stay public.

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Autoprobe.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Autoprobe.scala:99]: This is painful to do if we want an unbiased random, and to be honest we don't need it (the `Autoprobe` is unused or doesn't care about bias). I'd rather add it only if we ever need it...

File: eclair-core/src/main/scala/fr/acinq/eclair/package.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/package.scala:42]: Easy, I can export different traits for each purpose

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/package.scala]: Done in https://github.com/ACINQ/eclair/pull/1774/commits/0df44d9d5a3f6beac430640e213f8603185f57f0

File: eclair-core/src/main/scala/fr/acinq/eclair/package.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/package.scala]: Done in https://github.com/ACINQ/eclair/pull/1774/commits/0df44d9d5a3f6beac430640e213f8603185f57f0

File: eclair-core/src/main/scala/fr/acinq/eclair/package.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/package.scala]: This requires two byte array allocation for no reason, we can reuse at least one of them.
And xoring two `ByteVector` has to be less efficient than just iterating over the array to xor each byte individually.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/package.scala]: TBH I think this is really clear enough with the current code, and there's no risk of mutating the `ByteVector`, it's only applied after xor-ing everything.

**Review by pm47 — CHANGES_REQUESTED** [review]

File: eclair-core/src/main/scala/fr/acinq/eclair/package.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/package.scala:43]: That's right, I get turned off because IntelliJ prompts me to remove the parenthesis, but they do make sense.

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/WeakEntropyPool.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/WeakEntropyPool.scala]: Done in https://github.com/ACINQ/eclair/pull/1774/commits/54e961e5ba0f6be72d73211c47092989d4c02df9

File: eclair-core/src/main/scala/fr/acinq/eclair/package.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/package.scala]: Done in https://github.com/ACINQ/eclair/pull/1774/commits/54e961e5ba0f6be72d73211c47092989d4c02df9

File: eclair-core/src/main/scala/fr/acinq/eclair/package.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/package.scala]: Done in https://github.com/ACINQ/eclair/pull/1774/commits/1e1037f13dc7744449dda518e0cb3fa0c280e6e6
I kept it in a separate commit that does only this, since it's trivial to review but touches a lot of test files (which is painful to read).

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala
Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala]: LGTM

**Review by pm47 — DISMISSED** [review]
LGTM.

According to [this](https://security.stackexchange.com/a/47461), XORing two pseudorandom data sources cannot reduce randomness, as long as the two sources are unrelated to each other. This is the case here because we know exactly how the `WeakRandom` generator works.

**Comment by pm47** [comment]
Let's wait for a full review from @sstone before merging.

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/Random.scala
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/crypto/Random.scala:126]: Is it safe to use `bytes` here as both input and output to  `processBytes()` ?

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/Random.scala:126]: It should be how we do in-place stream encryption (without needing to allocate a new array for no reason).
But this is definitely an area where you should double-check that I didn't miss something!

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/Random.scala:126]: I can add a test vector to verify that this does what we expect.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/Random.scala:126]: I added a test case in https://github.com/ACINQ/eclair/pull/1774/commits/9fb8417d1b8b85a9c63904768f599a5ad82fb5e2 to prove this works as expected.

**Review by sstone — APPROVED** [review]

**Pull request merged**
